jerry-core/ecma/builtin-objects/ecma-builtin-global.c

```diff
@@ -361,30 +297,36 @@ ecma_builtin_global_object_decode_uri_helper (lit_utf8_byte_t *input_start_p, /*
       if (lit_is_code_point_utf16_high_surrogate (cp)
           || lit_is_code_point_utf16_low_surrogate (cp))
       {
-        ret_value = ecma_raise_uri_error (ECMA_ERR_MSG ("Invalid UTF8 codepoint."));
-        break;
+        ecma_stringbuilder_destroy (&builder);
+        return ecma_raise_uri_error (ECMA_ERR_MSG ("Invalid UTF8 codepoint."));
       }
 
-      output_char_p += lit_code_point_to_cesu8 (cp, output_char_p);
+      static lit_utf8_byte_t result_chars[6];
+      lit_utf8_size_t cp_size = lit_code_point_to_cesu8 (cp, result_chars);
+      ecma_stringbuilder_append_raw (&builder, result_chars, cp_size);
     }
   }
 
-  if (ecma_is_value_empty (ret_value))
-  {
-    JERRY_ASSERT (output_start_p + output_size == output_char_p);
+  ecma_string_t *str_p = ecma_stringbuilder_finalize (&builder);
+  lit_utf8_size_t result_size = ecma_string_get_size (str_p);
+  ecma_value_t ret_value = ECMA_VALUE_UNDEFINED;
 
-    if (lit_is_valid_cesu8_string (output_start_p, output_size))
-    {
-      ecma_string_t *output_string_p = ecma_new_ecma_string_from_utf8 (output_start_p, output_size);
-      ret_value = ecma_make_string_value (output_string_p);
-    }
-    else
-    {
-      ret_value = ecma_raise_uri_error (ECMA_ERR_MSG ("Invalid CESU8 string."));
-    }
+  JMEM_DEFINE_LOCAL_ARRAY (result_string,
```

Comment:
Instead of creating the string and copying into a fix sized array, you can use the stringbuilder internal data buffer.